_posts/2018-07-05-modelling-page-view-events-as-a-graph/index.md

```diff
@@ -0,0 +1,490 @@
+| Browser     | Browser                                             |
+| Device      | Device, OS                                          |
+| Session     | Session                                             |
+| Page        | Page, Document                                      |
+| User        | User                                                |
+| Network     | Network, Geo                                        |
+| Source      | Traffic source                                      |
+| Event       | Event metadata, Infrastructure metadata, Timestamps |
+
+We'll model the different timestamps as properties of the `Event` node. We can also model them as properties on the relationships within the event.
+
+You can find all of our proposed node schemas in [this GitHub repo][node-schemas].
+
+### Drawing the relationships
+
+The next step is to figure out all the relationships between our nodes. We already know what two of them will be:
+
+- `(Event)-[:NEXT]->(Event)`
+- `(Event)-[:HAS]->(Non-event node)`.
+
+We can consider all possible combinations of two nodes from the list above and see if any "natural" relationships exist between them. In that way, we can expand the list of relationships. To keep things simple for now, let's assume all constituent nodes are linked by a `[:WITH]` relationship, eg `(Application)<-[:WITH]-(Browser)`, `(Browser)<-[:WITH]-(Session)`.
+
+The proposed schemas for these relationships are in [this repo][relationship-schemas].
+
+### Composing the `page_view` schema
+
+We now have all the building blocks to compose a schema for the `page_view` event. You can find the schema [here][page-view-schema].
+
+If we use this model to represent a series of `page_view` events, we end up with a graph that looks something like this (most nodes are omitted, for clarity):
+
+![page-view-series][page-view-series]
+
+Event nodes (in grey) are connected via `NEXT` relationships. Each one of them has outgoing `HAS` relationships to constituent nodes, such as `Page` (magenta) and `User` (blue). In turn, the constituent nodes are linked via `WITH` relationships.
+
+Now we can easily query all the pages visited by a user, without having to go via the event node:
+
+```
+MATCH (u:User)-[:WITH]-(p:Page)
+WHERE u.user_id = 'someone@mail.com'
+RETURN p.page_url;
+```
+
+And we can also easily query the path a user has taken through the website:
+
+```
+MATCH p1 = (u:User)<-[:HAS]-(e:Event)
+WHERE u.user_id = 'someone@mail.com'
+WITH p1
+MATCH path = (e)-[:NEXT*1..5]-()
+RETURN path;
+```
+
+### Implementing the contract in code
+
+Now that we've schemaed everything, we can think about how we can implement this in actual code. A great thing about schemas is that the code falls naturally out of them. For example, here's a POJO representing the `Application` node along with a method that will create the node in the database (code to connect to Neo4j not shown):
+
+```Scala
+sealed trait Node
```

Comment:
I think python-like pseudo-code here would be more readable for analysts, who usually not very familiar with Scala (e.g. `Map[String, AnyRef]`, `Unit` etc are quite irrelevant details):

```python
class Node:
  def create(**options)

class Application(Node):
  def create(app_id, platform):
    statement = "MERGE (application:Application {app_id: $app_id, platform: $platform})"
    neo4j.execute(statement, {'platform': platform, 'app_id': app_id})

```